### PR TITLE
Implement updated spec rules related to CATs

### LINKED ISSUE #20746: NOC validity checking should ensure same CAT is not present with different versions
#### Problem
See https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5437 for the spec change.  We need to do that in the SDK.

 #### Proposed Solution
Add the check.

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- CAT identifiers could previously collide, which made their use ambiguous
  for access control
- CAT tag class did not enforce strongly that all CATs are at the front and
  this is not required for correctness
- CAT identifiers could be version 0 before

Based on https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5437
we need to fix this.

Fixes #20746

#### Change overview

- Checks CATs are valid when adding/updating NOCs
- Adds an `AreValid()` method to CATValues to ensure
  correctness

#### Testing
- Added unit tests for all new methods
- Cert tests still pass

### CODE REVIEW

**Comment by msandstedt** [comment]
This could be an opportunity to address this comment: https://github.com/project-chip/connectedhomeip/pull/20253/files#r913412111

**Review by emargolis — APPROVED** [review]

File: src/lib/core/CASEAuthTag.h [hunk 1]
```diff
@@ -101,9 +161,8 @@ struct CATValues
 
         for (auto cat : values)
         {
-            // All valid CAT values are always in the beginning of the array followed by kUndefinedCAT values.
-            ReturnErrorCodeIf(cat == kUndefinedCAT, false);
-            if (((cat & kTagIdentifierMask) == (subject & kTagIdentifierMask)) &&
+
+            if ((cat != kUndefinedCAT) && ((cat & kTagIdentifierMask) == (subject & kTagIdentifierMask)) &&
                 ((cat & kTagVersionMask) >= (subject & kTagVersionMask)))
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/lib/core/CASEAuthTag.h]: Applied by paraphrase

File: src/lib/core/CASEAuthTag.h [hunk 2]
```diff
@@ -73,6 +98,41 @@ struct CATValues
...
+                return false;
+            }
+            // Identifiers cannot collide in set (there cannot be more than 1 version of an identifier)
+            for (size_t other_idx = 0; other_idx < size(); ++other_idx)
+            {
+                if (idx == other_idx)
+                {
+                    continue;
+                }
+
```
Comment (bzbarsky-apple) [inline src/lib/core/CASEAuthTag.h:123]:
```suggestion

                if (values[other_idx] == kUndefinedCAT)
                {
                    continue;
                }

```

File: src/lib/core/CASEAuthTag.h [hunk 3]
```diff
@@ -98,13 +158,12 @@ struct CATValues
...
+        CASEAuthTag candidate = CASEAuthTagFromNodeId(subject);
 
         for (auto cat : values)
         {
-            // All valid CAT values are always in the beginning of the array followed by kUndefinedCAT values.
-            ReturnErrorCodeIf(cat == kUndefinedCAT, false);
-            if (((cat & kTagIdentifierMask) == (subject & kTagIdentifierMask)) &&
-                ((cat & kTagVersionMask) >= (subject & kTagVersionMask)))
+            if ((cat != kUndefinedCAT) && (GetCASEAuthTagIdentifier(cat) == GetCASEAuthTagIdentifier(candidate)) &&
+                (GetCASEAuthTagVersion(candidate) > 0) && (GetCASEAuthTagVersion(cat) >= GetCASEAuthTagVersion(candidate)))
```
Comment (bzbarsky-apple) [inline src/lib/core/CASEAuthTag.h:166]: Pre-existing, but this looks wrong to me.  Shouldn't this be checking that the _candidate_ version is at least the version of `cat`?  Certainly the spec has:
```
     (get_cat_id(acl_subject) == get_cat_id(isd_subject)) and
     (get_cat_version(isd_subject) >= get_cat_version(acl_subject))
```
to be considered a match.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/core/CASEAuthTag.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/lib/core/CASEAuthTag.h:166]: Per Slack discussion, this code is correct.  "subject" here is an _acl_ subject, not an ISD subject, and "cat" is one of the ISD subjects.

**Comment by tcarmelveilleux** [comment]
> This could be an opportunity to address this comment: https://github.com/project-chip/connectedhomeip/pull/20253/files#r913412111

Done

**Review by msandstedt — APPROVED** [review]

**Pull request merged**